test/fuzz/fuzzer_pass_donate_modules_test.cpp

```diff
@@ -1678,6 +1680,119 @@ TEST(FuzzerPassDonateModulesTest, Miscellaneous1) {
+               OpSource ESSL 310
+          %2 = OpTypeVoid
+          %3 = OpTypeFunction %2
+          %4 = OpFunction %2 None %3
+          %5 = OpLabel
+               OpReturn
+               OpFunctionEnd
+  )";
+
+  const auto env = SPV_ENV_UNIVERSAL_1_3;
+  const auto consumer = nullptr;
+  const auto recipient_context =
+      BuildModule(env, consumer, recipient_shader, kFuzzAssembleOption);
+  ASSERT_TRUE(IsValid(env, recipient_context.get()));
+
+  const auto donor_context =
+      BuildModule(env, consumer, donor_shader, kFuzzAssembleOption);
+  ASSERT_TRUE(IsValid(env, donor_context.get()));
+
+  FactManager fact_manager;
+  spvtools::ValidatorOptions validator_options;
+  TransformationContext transformation_context(&fact_manager,
+                                               validator_options);
+
+  PseudoRandomGenerator prng(0);
+  FuzzerContext fuzzer_context(&prng, 100);
+  protobufs::TransformationSequence transformation_sequence;
+
+  FuzzerPassDonateModules fuzzer_pass(recipient_context.get(),
+                                      &transformation_context, &fuzzer_context,
+                                      &transformation_sequence, {});
+
+  fuzzer_pass.DonateSingleModule(donor_context.get(), false);
+
+  // We just check that the result is valid.  Checking to what it should be
+  // exactly equal to would be very fragile.
+  ASSERT_TRUE(IsValid(env, recipient_context.get()));
+
+  // Check that OpSpecConstant* instructions were correctly donated to the
+  // module.
+  bool checks[4] = {};
+  for (const auto* instr : recipient_context->GetConstants()) {
+    switch (instr->opcode()) {
+      case SpvOpConstantFalse: {
+        // The module has OpConstantFalse and we have already checked that the
+        // module is valid, so this check is accepted.
+        checks[0] = true;
+      } break;
+      case SpvOpConstantTrue: {
+        // OpConstantTrue is present and the module is valid - check.
+        checks[1] = true;
+      } break;
+      case SpvOpConstant: {
+        const auto* type =
+            recipient_context->get_type_mgr()->GetType(instr->type_id());
+
+        spvtools::opt::analysis::Integer int_type(32, true);
+        ASSERT_TRUE(type && type->IsSame(&int_type));
+        ASSERT_TRUE(instr->GetSingleWordInOperand(0) == 2);
+        checks[2] = true;
+      } break;
+      case SpvOpConstantComposite: {
+        const auto* type =
+            recipient_context->get_type_mgr()->GetType(instr->type_id());
+
+        // Check that struct type corresponds to the one from the donor shader.
+        spvtools::opt::analysis::Integer int_type(32, true);
+        spvtools::opt::analysis::Bool bool_type;
+        spvtools::opt::analysis::Struct struct_type(
+            {&bool_type, &bool_type, &int_type});
+        ASSERT_TRUE(type && type->IsSame(&struct_type));
+
+        checks[3] = true;
+      } break;
+      default:
+        // We are not interested in other instructions
+        break;
+    }
```

Comment:
Not sure how we should check `OpSpecConstantOp`. Parser always fails when I try to include it to the donor module. It seems I'm missing something here.

Reply:
Here is an example that uses OpConstantOp:

```
               OpCapability Shader
          %1 = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint Fragment %4 "main"
               OpExecutionMode %4 OriginUpperLeft
               OpSource ESSL 320
               OpName %4 "main"
        %100 = OpTypeInt 32 0
        %101 = OpSpecConstant %100 1
        %102 = OpSpecConstant %100 2
        %103 = OpSpecConstantOp %102 IAdd %101 %102
          %2 = OpTypeVoid
          %3 = OpTypeFunction %2
          %4 = OpFunction %2 None %3
          %5 = OpLabel
               OpReturn
               OpFunctionEnd
```

It turns out that you just put e.g. IAdd, not OpIAdd or 128, for the operand literal.  I figured this out by searching for OpSpecConsantOp in the code base and found something in assembly_builder_test.cpp that uses it.

Reply:
See my comment in the main conversation on this.

Reply:
Fixed in 1ab2f64.